### PR TITLE
Icinga DB: use one Redis connection per config object type

### PR DESCRIPTION
**Pull request by Al2Klimov**
... to double the initial dump performance.

### CODE REVIEW

**Comment by Al2Klimov** [comment]
## Before/after ([bench-idbr](https://github.com/Al2Klimov/bench-idbr))

```
0.02  wip: [*]
0.05  done: [externalcommandlistener dbconnection scheduleddowntime logger]
0.09  done: [checkercomponent]
1.92  done: [icingaapplication dependency command icingadb idomysqlconnection idopgsqlconnection influxdbwriter livestatuslistener sysloglogger]
2.02  done: [apilistener apiuser filelogger]
2.05  done: [application timeperiod gelfwriter]
2.09  done: [user graphitewriter]
2.25  done: [usergroup]
2.69  done: [downtime elasticsearchwriter]
7.12  done: [endpoint]
7.19  done: [eventcommand]
8.72  done: [zone]
9.75  done: [checkcommand]
9.89  done: [checkresultreader]
10.12  done: [checkable]
26.02  done: [comment]
26.05  done: [compatlogger configobject customvarobject]
50.25  done: [host]
50.29  done: [hostgroup]
70.99  done: [notification]
71.09  done: [notificationcommand]
71.12  done: [notificationcomponent opentsdbwriter perfdatawriter]
96.12  done: [service servicegroup statusdatawriter streamlogger *]
0.02  wip: [*]
0.02  done: [dbconnection checkercomponent apilistener externalcommandlistener logger sysloglogger icingaapplication scheduleddowntime dependency apiuser filelogger command gelfwriter timeperiod]
2.29  done: [icingadb application graphitewriter user]
2.42  done: [idomysqlconnection usergroup]
2.45  done: [idopgsqlconnection downtime influxdbwriter]
2.49  done: [elasticsearchwriter]
2.55  done: [livestatuslistener]
4.95  done: [endpoint]
4.99  done: [eventcommand]
5.32  done: [zone]
5.62  done: [checkcommand]
5.65  done: [checkresultreader]
5.69  done: [checkable]
11.92  done: [comment]
11.95  done: [compatlogger]
11.99  done: [configobject customvarobject]
13.92  done: [host]
14.29  done: [hostgroup]
23.25  done: [notification]
23.29  done: [notificationcommand notificationcomponent opentsdbwriter perfdatawriter]
37.22  done: [service]
38.45  done: [servicegroup statusdatawriter streamlogger *]
```

**Comment by Al2Klimov** [comment]
> actually modern versions of Redis use threads for different things

– https://redis.io/topics/benchmarks

**Comment by Al2Klimov** [comment]
I've tried to test the _a lot of runtime updates just after the "finished dump"_ problem, but it seems to be gone independently of this patch.

File: lib/icingadb/icingadb-objects.cpp [hunk 1]
```diff
@@ -115,6 +115,17 @@ void IcingaDB::ConfigStaticInitialize()
...
 
+/**
+ * Puts a no-op command with a result at the end of the queue and wait for the result,
+ * i.e. for everything enqueued to be processed by the server.
+ *
+ * @ingroup icingadb
+ */
+static void SyncRedisConnection(const RedisConnection::Ptr& rCon)
+{
+	rCon->GetResultOfQuery({"PING"}, (RedisConnection::QueryPriority)42);
```
Comment (julianbrost) [inline lib/icingadb/icingadb-objects.cpp:126]: When we actually consider merging this, this should get a proper priority value.

**Review by julianbrost — COMMENTED** [review]
Looks fine from what it does in general (haven't tested it myself though).

Is there any reason for storing these additional connections in a class member compared to just opening them when doing the initial sync and closing them afterwards? I don't think they will ever be reused anyways.

Something else to consider: it still looks like the total sync time is dominated by the services. Maybe this could be improved further by distributing each object type across a shared connection pool? Might or might not give an improvement depending on the current bottleneck.

File: lib/icingadb/icingadb-objects.cpp (hunk 1 above)
Comment (Al2Klimov) [reply lib/icingadb/icingadb-objects.cpp]: ... or the just done refactoring.

**Comment by Al2Klimov** [comment]
> I don't think they will ever be reused anyways.

I'm very surprised to hear this... from _you_. (#8604)

**Comment by julianbrost** [comment]
So if you manage to reset the main Redis connection but keep the other ones alive, then yes, they would get reused. But I think most common cases for the main connection to reset would also reset all other connections to the same Redis server.

**Comment by Al2Klimov** [comment]
## Blocked by

* [ ] #8641

**Comment by Al2Klimov** [comment]
Re-tested. Actually boosts by 15% if the Redis is in the same DC.

**Comment by Al2Klimov** [comment]
> Is there any reason for storing these additional connections in a class member compared to just opening them when doing the initial sync and closing them afterwards? I don't think they will ever be reused anyways.
> 
> Something else to consider: it still looks like the total sync time is dominated by the services. Maybe this could be improved further by distributing each object type across a shared connection pool? Might or might not give an improvement depending on the current bottleneck.

1. Due to the design of the connection objects it's better to just keep them around.
2. I opt for not too many connections.

**Comment by julianbrost** [comment]
I think in the meantime there were quite some other changes on how and what we write to Redis. Do we still think we need this so that it makes sense to resolve the conflicts and reevaluate?

**Review by julianbrost — COMMENTED** [review]
Got a performance increase of about 4% when running Icinga 2 and Redis on the same host.

File: lib/icingadb/icingadb-objects.cpp [hunk 2]
```diff
@@ -55,6 +55,27 @@ return id
...
+		Notification::TypeInstance,
+		NotificationCommand::TypeInstance,
+		Service::TypeInstance,
+		ServiceGroup::TypeInstance,
+		TimePeriod::TypeInstance,
+		User::TypeInstance,
+		UserGroup::TypeInstance,
+		Zone::TypeInstance
+	};
+}
```
[the remark this answers is not in the document]
Comment (julianbrost) [reply lib/icingadb/icingadb-objects.cpp:77]: Ok, turns out this doesn't work as `T::TypeInstance` is initialized too late for this.

**Review by julianbrost — APPROVED** [review]
Added a small change to not capture everything in a Lambda. Good to go from my side then.

Since I changed the PR, please have another look @N-o-X

**Review by N-o-X — APPROVED** [review]
LGTM 👍

**Pull request merged**